core/lib/src/shutdown.rs

```diff
@@ -0,0 +1,92 @@
+use futures::{channel::mpsc, stream::StreamExt};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Wrapper around an mpsc channel.
+#[derive(Debug)]
+crate struct Shutdown {
+    sender: mpsc::Sender<()>,
+    crate receiver: Option<Receiver>,
+}
+
+impl Shutdown {
+    /// Create a `Shutdown`.
+    #[inline]
+    crate fn new() -> Self {
+        let (sender, receiver) = mpsc::channel(1);
+        Self {
+            sender,
+            receiver: Some(Receiver { receiver }),
+        }
+    }
+
+    /// Get a handle that can be used to call `.shutdown()`.
+    #[inline]
+    crate fn get_handle(&self) -> ShutdownHandle<'_> {
+        ShutdownHandle {
+            sender: &self.sender,
+        }
+    }
+}
+
+/// # Example
+///
+/// ```rust
+/// # #![feature(proc_macro_hygiene, async_await)]
+/// # #[macro_use] extern crate rocket;
+/// #
+/// use rocket::shutdown::ShutdownHandle;
+///
+/// #[get("/shutdown")]
+/// fn shutdown(handle: &ShutdownHandle) -> &'static str {
+///     handle.shutdown();
+///     "Shutting down..."
+/// }
+///
+/// fn main() {
+///     # if false {
+///     rocket::ignite()
+///         .mount("/", routes![shutdown])
+///         .launch();
+///     # }
+/// }
+/// ```
+#[derive(Debug, Clone, Copy)]
+pub struct ShutdownHandle<'a> {
+    sender: &'a mpsc::Sender<()>,
+}
+
+impl ShutdownHandle<'_> {
+    /// Notify Rocket to shut down gracefully.
+    ///
+    /// This method should only need to be called once.
+    #[inline]
+    pub fn shutdown(&self) {
+        // Clone to avoid requiring `self` to be mutable.
+        self.sender
+            .clone()
+            .try_send(())
+            .expect("Error sending shutdown signal")
+    }
+}
+
+/// Thin wrapper around an `mpsc::Receiver`,
+/// implementing `Future`, which waits for the first value.
+#[derive(Debug)]
+crate struct Receiver {
+    receiver: mpsc::Receiver<()>,
+}
+
+impl Future for Receiver {
+    type Output = ();
+
+    #[inline]
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.receiver
+            .poll_next_unpin(cx)
+            .map(|opt| opt.unwrap_or(()))
```

Comment:
This is a bit weird. When could the future need to be polled after all senders are already gone?

Reply:
IIRC I had to do this because it's a channel, so there are error types associated. In this situation, I don't believe it's possible to poll after all senders are `drop`ped.